File: aznavrail-react/src/util/AzRailLayoutHelper.ts
import { AzDockingSide, AzOrientation, AzVisualSide } from '../types';

/** Resolved layout parameters produced by `AzRailLayoutHelper.calculateLayout`. */
export interface RailLayoutConfig {
  /** The physical screen edge where the rail appears after rotation is applied. */
  visualSide: AzVisualSide;
  /** Whether items in the rail flow vertically or horizontally after rotation. */
  orientation: AzOrientation;
  /** Alignment corner used when placing the rail in landscape/rotated layouts. */
  alignment: 'TopStart' | 'TopEnd' | 'BottomStart' | 'BottomEnd';
  /** When true, items within the rail should be rendered in reverse order. */
  reverseLayout: boolean;
}

/** Utility class that maps a `dockingSide` + device rotation to resolved visual-layout parameters. */
export class AzRailLayoutHelper {
  /**
   * Calculates the visual layout of the rail given device rotation and docking preferences.
   * @param dockingSide - The logical docking side chosen by the user.
   * @param rotation - Current device rotation in degrees (0, 90, 180, or 270).
   * @param usePhysicalDocking - When true, the rail follows the physical edge across rotations.
   */
  static calculateLayout(
    dockingSide: AzDockingSide,
    rotation: number, // 0, 90, 180, 270
    usePhysicalDocking: boolean
  ): RailLayoutConfig {
    let visualSide: AzVisualSide;

    if (usePhysicalDocking) {
      if (dockingSide === AzDockingSide.LEFT) {
        switch (rotation) {
          case 0: visualSide = AzVisualSide.LEFT; break;
          case 90: visualSide = AzVisualSide.BOTTOM; break;
          case 180: visualSide = AzVisualSide.RIGHT; break;
          case 270: visualSide = AzVisualSide.TOP; break;
          default: visualSide = AzVisualSide.LEFT;
        }
      } else {
        switch (rotation) {
          case 0: visualSide = AzVisualSide.RIGHT; break;
          case 90: visualSide = AzVisualSide.TOP; break;
          case 180: visualSide = AzVisualSide.LEFT; break;
          case 270: visualSide = AzVisualSide.BOTTOM; break;
          default: visualSide = AzVisualSide.RIGHT;
        }
      }
    } else {
      visualSide = dockingSide === AzDockingSide.LEFT ? AzVisualSide.LEFT : AzVisualSide.RIGHT;
    }

    const orientation = (visualSide === AzVisualSide.TOP || visualSide === AzVisualSide.BOTTOM)
      ? AzOrientation.HORIZONTAL
      : AzOrientation.VERTICAL;

    let reverseLayout = false;
    if (usePhysicalDocking) {
      if (dockingSide === AzDockingSide.LEFT) {
        reverseLayout = (rotation === 180 || rotation === 270);
      } else {
        reverseLayout = (rotation === 180 || rotation === 90);
      }
    }

    let alignment: 'TopStart' | 'TopEnd' | 'BottomStart' | 'BottomEnd';
    switch (visualSide) {
      case AzVisualSide.LEFT: alignment = 'TopStart'; break;
      case AzVisualSide.RIGHT: alignment = 'TopEnd'; break;
      case AzVisualSide.TOP: alignment = 'TopStart'; break;
      case AzVisualSide.BOTTOM: alignment = 'BottomStart'; break;
    }

    return { visualSide, orientation, alignment, reverseLayout };
  }
}

File: aznavrail-cmp/src/commonMain/kotlin/com/hereliesaz/aznavrail/internal/AzRailLayoutHelper.kt
package com.hereliesaz.aznavrail.internal

import androidx.compose.ui.Alignment
import com.hereliesaz.aznavrail.model.AzDockingSide
import com.hereliesaz.aznavrail.model.AzOrientation

/**
 * Enumeration representing the visual side of the screen where the rail is docked.
 * This takes into account physical device rotation.
 */
enum class AzVisualSide { LEFT, RIGHT, TOP, BOTTOM }

/**
 * Data class holding the calculated layout configuration for the rail.
 *
 * @param visualSide The effective visual side of the rail.
 * @param orientation The orientation of the rail items.
 * @param alignment The alignment of the rail within the root container.
 * @param reverseLayout Whether the item order should be reversed (e.g., for bottom docking).
 */
data class RailLayoutConfig(
    val visualSide: AzVisualSide,
    val orientation: AzOrientation,
    val alignment: Alignment,
    val reverseLayout: Boolean,
)

/**
 * Helper object for calculating rail layout based on docking side and device rotation.
 *
 * Port note vs the Android sibling: `rotation` is a Float (degrees clockwise from portrait — 0f,
 * 90f, 180f, or 270f) rather than an Int matching `android.view.Surface.ROTATION_*`. Callers
 * supply the value via [rememberDeviceRotationDegrees]; on non-Android targets the helper always
 * receives 0f (no physical rotation compensation), matching the desktop/iOS/wasmJs UX where the
 * window is portrait-orientation-only or the concept doesn't apply.
 */
internal object AzRailLayoutHelper {
    /**
     * Calculates the layout configuration.
     *
     * @param dockingSide The configured logical docking side.
     * @param rotation The current display rotation, in degrees clockwise from portrait
     *   (0f / 90f / 180f / 270f).
     * @param usePhysicalDocking Whether to apply physical docking logic (adapting to rotation).
     * @return The calculated [RailLayoutConfig].
     */
    fun calculateLayout(
        dockingSide: AzDockingSide,
        rotation: Float,
        usePhysicalDocking: Boolean,
    ): RailLayoutConfig {
        val visualSide = if (usePhysicalDocking) {
            when (dockingSide) {
                AzDockingSide.LEFT -> when (rotation) {
                    0f -> AzVisualSide.LEFT
                    90f -> AzVisualSide.BOTTOM
                    180f -> AzVisualSide.RIGHT
                    270f -> AzVisualSide.TOP
                    else -> AzVisualSide.LEFT
                }
                AzDockingSide.RIGHT -> when (rotation) {
                    0f -> AzVisualSide.RIGHT
                    90f -> AzVisualSide.TOP
                    180f -> AzVisualSide.LEFT
                    270f -> AzVisualSide.BOTTOM
                    else -> AzVisualSide.RIGHT
                }
            }
        } else {
            // Stick to view side (Classic behavior)
            if (dockingSide == AzDockingSide.LEFT) AzVisualSide.LEFT else AzVisualSide.RIGHT
        }

        val orientation = if (visualSide == AzVisualSide.TOP || visualSide == AzVisualSide.BOTTOM)
            AzOrientation.Horizontal
        else
            AzOrientation.Vertical

        val reverseLayout = if (usePhysicalDocking) {
            when (dockingSide) {
                AzDockingSide.LEFT -> (rotation == 180f || rotation == 270f)
                AzDockingSide.RIGHT -> (rotation == 180f || rotation == 90f)
            }
        } else {
            false
        }

        val alignment = when (visualSide) {
            AzVisualSide.LEFT -> Alignment.TopStart
            AzVisualSide.RIGHT -> Alignment.TopEnd
            AzVisualSide.TOP -> Alignment.TopStart
            AzVisualSide.BOTTOM -> Alignment.BottomStart
        }

        return RailLayoutConfig(visualSide, orientation, alignment, reverseLayout)
    }
}
